File: Java/problem1160.java
class Solution {
    private int[] charNum;

    private boolean fromWords(String str) {
        int[] tmp = new int[26];
        for (int i = 0; i < str.length(); ++i) {
            int item = (int)str.charAt(i) - (int)'a';
            if (++tmp[item] > charNum[item]) {
                return false;
            }
        }
        return true;
    }

    public int countCharacters(String[] words, String chars) {
        charNum = new int[26];
        for (int i = 0; i < chars.length(); ++i) {
            char item = chars.charAt(i);
            ++charNum[(int)item - (int)'a'];
        }
        int sumLen = 0;
        for (String w : words) {
            if (fromWords(w)) {
                sumLen += w.length();
            }
        }
        return sumLen;
    }
}

File: python/problem1160.py
class Solution:
    def countCharacters(self, words: List[str], chars: str) -> int:
        charNum = [0 for _ in range(26)]
        for i in range(len(chars)):             #记录字母表中各个字母出现次数
            charNum[ord(chars[i]) - 97] += 1
        
        def fromWords(word):                    #判断某个单词能否由字母表中的字母拼成
            tmp = [0 for _ in range(26)]
            for i in range(len(word)):
                idx = ord(word[i]) - 97
                if tmp[idx] >= charNum[idx]:    #超过次数的就失败了
                    return False
                tmp[idx] += 1
            return True
        
        sumLen = 0
        for i in range(len(words)):
            if (fromWords(words[i])):
                sumLen += len(words[i])
        return sumLen
